File: epi_judge_python_solutions/closest_int_same_weight.py
from test_framework import generic_test


def closest_int_same_bit_count(x: int) -> int:

    num_unsigned_bits = 64
    for i in range(num_unsigned_bits - 1):
        if (x >> i) & 1 != (x >> (i + 1)) & 1:
            x ^= (1 << i) | (1 << (i + 1))  # Swaps bit-i and bit-(i + 1).
            return x

    # Raise error if all bits of x are 0 or 1.
    raise ValueError('All bits are 0 or 1')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('closest_int_same_weight.py',
                                       'closest_int_same_weight.tsv',
                                       closest_int_same_bit_count))

File: epi_judge_java_solutions/epi/ClosestIntSameWeight.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class ClosestIntSameWeight {

  @EpiTest(testDataFile = "closest_int_same_weight.tsv")
  public static long closestIntSameBitCount(long x) {

    final int NUM_UNSIGNED_BITS = 63;
    // x is assumed to be non-negative so we know the leading bit is 0. We
    // restrict to our attention to 63 LSBs.
    for (int i = 0; i < NUM_UNSIGNED_BITS - 1; ++i) {
      if ((((x >>> i) & 1) != ((x >>> (i + 1)) & 1))) {
        x ^= (1L << i) | (1L << (i + 1)); // Swaps bit-i and bit-(i + 1).
        return x;
      }
    }

    // Throw error if all bits of x are 0 or 1.
    throw new IllegalArgumentException("All bits are 0 or 1");
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "ClosestIntSameWeight.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
